util/resolver/resolver.go

```diff
@@ -232,9 +238,15 @@ func (f *httpFallback) RoundTrip(r *http.Request) (*http.Response, error) {
 	plainHTTPRequest := *r
 	plainHTTPRequest.URL = &plainHTTPUrl
 
+	f.hostMut.Lock()
+	updateBody := false
 	if f.host != r.URL.Host {
 		f.host = r.URL.Host
+		updateBody = true
+	}
+	f.hostMut.Unlock()
 
+	if updateBody {
```

Comment:
Its probably better to just check `fhost != r.URL.host` here, relying on the check above it is possible that f.host changed but the body still needs update.

My recommendation would be check fhost here first, then lock, check, update after falling into this block. This also saves one unneeded lock in the fallback case.

Reply:
Ah good catch, fixed. Also added some comments to make it more readable.